depthai_ros_driver/include/depthai_ros_driver/impl/ros_agnostic.hpp

```diff
@@ -0,0 +1,249 @@
+  }
+
+private:
+  /// @brief Create ros publisher
+  template<class Msg>
+  void create_publisher(
+    ROSNodeHandle node_handle,
+    const std::string& topic_name,
+    const uint32_t queue_size)
+  {
+    #if defined(USE_ROS2)
+    // using qos: KeepLast(QueueSize)
+    _ptr = node_handle->create_publisher<Msg>(topic_name, queue_size);
+    #else
+    _pub = node_handle->advertise<Msg>(topic_name, queue_size);
+    #endif
+  }
+
+  #if defined(USE_ROS2)
+  std::shared_ptr<void> _ptr;
+  #else
+  ros::Publisher _pub;
+  #endif
+
+  friend class NodeInterface;
+};
+
+//==============================================================================
+class Subscription
+{
+private:
+  /// @brief Create ros subscription
+  template<class Msg, typename Callback>
+  void create_subscription(
+    ROSNodeHandle node_handle,
+    const std::string& topic_name,
+    const uint32_t queue_size,
+    const Callback& callback)
+  {
+    #if defined(USE_ROS2)
+    // using qos: KeepLast(QueueSize)
+    _ptr = node_handle->create_subscription<Msg>(topic_name, queue_size,
+        callback);
+    #else
+    _sub = node_handle->subscribe<Msg>(topic_name, queue_size, callback);
+    #endif
+  }
+
+  #if defined(USE_ROS2)
+  std::shared_ptr<void> _ptr;
+  #else
+  ros::Subscriber _sub;
+  #endif
+
+  friend class NodeInterface;
+};
+
+//==============================================================================
+class Service
+{
+private:
+  /// @brief Create ros service
+  template<class Msg, typename Callback>
+  void create_service(
+    ROSNodeHandle node_handle,
+    const std::string& srv_name,
+    const Callback& callback)
+  {
+    #if defined(USE_ROS2)
+    _ptr = node_handle->create_service<Msg>(srv_name, callback);
+    #else
+    using RequestMsg = typename Msg::Request;
+    using ResponseMsg = typename Msg::Response;
+    _srv = node_handle->advertiseService<RequestMsg, ResponseMsg>(
+      srv_name,
+      [callback = std::move(callback)](const RequestMsg& req, ResponseMsg& res)
+      {
+        // internally convert ref to ptr
+        auto res_ptr = std::make_shared<ResponseMsg>();
+        callback(std::make_shared<RequestMsg>(req), res_ptr);
+        res = *res_ptr;
+        return true;
+      });
+    #endif
+  }
+
+  #if defined(USE_ROS2)
+  std::shared_ptr<void> _ptr;
+  #else
+  ros::ServiceServer _srv;
+  #endif
+
+  friend class NodeInterface;
+};
+
+//==============================================================================
+class Timer
+{
+private:
+  /// @brief Create ros timer
+  template<typename Callback>
+  void create_timer(
+    ROSNodeHandle node_handle,
+    const double period_sec,
+    const Callback& callback)
+  {
+    #if defined(USE_ROS2)
+    const auto period = std::chrono::duration<double>(period_sec);
+    _timer = node_handle->create_wall_timer(period, callback);
+    #else
+    _timer = node_handle->createTimer(ros::Duration(period_sec),
+        [callback = std::move(callback)](const ros::TimerEvent&)
```

Comment:
callback is a const reference. Would a cast to `const Callback&&` give any benefits?

Reply:
without the move in the lambda capture, the callback function can get invalid. Adding this solved the seg fault prob.